### PR TITLE
[SPARK-14353] Dataset Time Window `window` API for Python, and SQL

### PR DESCRIPTION
**Pull request by brkyvz**
## What changes were proposed in this pull request?

The `window` function was added to Dataset with [this PR](https://github.com/apache/spark/pull/12008).
This PR adds the Python, and SQL, API for this function.

With this PR, SQL, Java, and Scala will share the same APIs as in users can use:
- `window(timeColumn, windowDuration)` 
- `window(timeColumn, windowDuration, slideDuration)` 
- `window(timeColumn, windowDuration, slideDuration, startTime)`

In Python, users can access all APIs above, but in addition they can do
- In Python:
  `window(timeColumn, windowDuration, startTime=...)`

that is, they can provide the startTime without providing the `slideDuration`. In this case, we will generate tumbling windows.
## How was this patch tested?

Unit tests + manual tests

### CODE REVIEW

**Comment by rxin** [comment]
Can we leave all the streaming stuff out of R for now? I don't know whether it makes sense to provide this functionality in R, and even if we do, the current API might change a lot, so having it in fewer languages would make it easier to change.

**Comment by rxin** [comment]
Ah OK this "window" thing is not streaming specific. I'd still consider breaking them into two separate prs so we can get the python thing in sooner.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala [hunk 1]
```diff
@@ -365,20 +365,29 @@ abstract class TreeNode[BaseType <: TreeNode[BaseType]] extends Product {
...
     if (ctors.isEmpty) {
       sys.error(s"No valid constructor for $nodeName")
     }
-    val defaultCtor = ctors.maxBy(_.getParameterTypes.size)
+    val allArgs: Array[AnyRef] = if (otherCopyArgs.isEmpty) {
+      newArgs
+    } else {
+      newArgs ++ otherCopyArgs
+    }
+    val defaultCtor = ctors.find { ctor =>
```
Comment (brkyvz) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala]: we need a smarter way to pick the constructor, or else the two constructors for `TimeWindow` cause issues depending on Java version

File: python/pyspark/sql/functions.py [hunk 2]
```diff
@@ -1053,6 +1053,44 @@ def to_utc_timestamp(timestamp, tz):
...
+    and 'end'.
+
+    >>> df = sqlContext.createDataFrame([("2016-03-11 09:00:07", 1)]).toDF("date", "val")
+    >>> w = df.groupBy(window("date", "5 seconds")).agg(sum("val").alias("sum"))
+    >>> w.select(w.window.start.cast("string"), w.window.end.cast("string"), "sum").collect()
+    [Row(start="2016-03-11 09:00:05, end=2016-03-11 09:00:10, sum=1)]
+    """
+    sc = SparkContext._active_spark_context
+    time_col = _to_java_column(col)
+    if slideDuration and startTime:
```
Comment (davies) [inline python/pyspark/sql/functions.py]: It's better to check the type of slideDuration and startTime here, raise an Python TypeError exception if there are not str.

File: python/pyspark/sql/functions.py [hunk 3]
```diff
@@ -1053,6 +1053,44 @@ def to_utc_timestamp(timestamp, tz):
...
+    window intervals. For example, in order to have hourly tumbling windows that start 15 minutes
+    past the hour, e.g. 12:15-13:15, 13:15-14:15... provide `startTime` as `15 minutes`.
+
+    The output column will be a struct called 'window' by default with the nested columns 'start'
+    and 'end'.
+
+    >>> df = sqlContext.createDataFrame([("2016-03-11 09:00:07", 1)]).toDF("date", "val")
+    >>> w = df.groupBy(window("date", "5 seconds")).agg(sum("val").alias("sum"))
+    >>> w.select(w.window.start.cast("string"), w.window.end.cast("string"), "sum").collect()
+    [Row(start="2016-03-11 09:00:05, end=2016-03-11 09:00:10, sum=1)]
```
Comment (davies) [inline python/pyspark/sql/functions.py]:
```
Row(start=u"2016-03-11 09:00:05", end=u"2016-03-11 09:00:10", sum=1)]
```
You may also need  @ignore_unicode_prefix

File: python/pyspark/sql/functions.py [hunk 4]
```diff
@@ -1053,6 +1053,44 @@ def to_utc_timestamp(timestamp, tz):
...
+    Durations are provided as strings, e.g. '1 second', '1 day 12 hours', '2 minutes'. Valid
+    interval strings are 'week', 'day', 'hour', 'minute', 'second', 'millisecond', 'microsecond'.
+    If the `slideDuration` is not provided, the windows will be tumbling windows.
+
+    The startTime is the offset with respect to 1970-01-01 00:00:00 UTC with which to start
+    window intervals. For example, in order to have hourly tumbling windows that start 15 minutes
+    past the hour, e.g. 12:15-13:15, 13:15-14:15... provide `startTime` as `15 minutes`.
+
+    The output column will be a struct called 'window' by default with the nested columns 'start'
+    and 'end'.
```
Comment (davies) [inline python/pyspark/sql/functions.py]: What's the type of start and end? TimestampType?

**Comment by brkyvz** [comment]
@davies Addressed your comments

**Comment by davies** [comment]
LGTM, merging this into master, thanks!

**Pull request closed without merging**